xapian-applications/omega/handler_libarchive.cc

```diff
@@ -31,8 +34,221 @@
+		}
+	    } else if (pathname == "docProps/core.xml") {
+		// docProps/core.xml stores meta data
+		total = archive_entry_size(entry);
+		metadata.resize(total);
+		size = archive_read_data(archive_obj, &metadata[0], total);
+		if (size > 0) {
+		    metaxml = true;
+		    metadata.resize(size);
+		}
+	    }
+	}
+    } else if (startswith(tail, "spreadsheetml.")) {
+	string shared_strings, sheets;
+	while (archive_read_next_header(archive_obj, &entry) == ARCHIVE_OK) {
+	    string pathname = archive_entry_pathname(entry);
+	    if (pathname == "xl/styles.xml" ||
+		pathname == "xl/workbook.xml" ||
+		pathname == "xl/sharedStrings.xml") {
+		auto i = shared_strings.size();
+		total = archive_entry_size(entry);
+		shared_strings.resize(i + total);
+		size = archive_read_data(archive_obj, &shared_strings[i],
+					 total);
+
+		if (size > 0)
+		    shared_strings.resize(i + size);
+		else
+		    shared_strings.resize(i);
+	    } else if (startswith(pathname, "xl/worksheets/sheet")) {
+		auto i = sheets.size();
+		total = archive_entry_size(entry);
+		sheets.resize(i + total);
+		size = archive_read_data(archive_obj, &sheets[i], total);
+
+		if (size <= 0) {
+		    error = "Libarchive was not able to extract data from " +
+			    pathname;
+		    return false;
+		}
+		sheets.resize(i + size);
+	    } else if (pathname == "docProps/core.xml") {
+		total = archive_entry_size(entry);
+		metadata.resize(total);
+		size = archive_read_data(archive_obj, &metadata[0], total);
+		if (size > 0) {
+		    metaxml = true;
+		    metadata.resize(size);
+		}
+	    }
+	}
+	content = shared_strings + sheets;
+    } else if (startswith(tail, "presentationml.")) {
+	msxml = true;
+	while (archive_read_next_header(archive_obj, &entry) == ARCHIVE_OK) {
+	    string pathname = archive_entry_pathname(entry);
+	    if (startswith(pathname, "ppt/slides/slide") ||
+		startswith(pathname, "ppt/notesSlides/notesSlide") ||
+		startswith(pathname, "ppt/comments/comment")) {
+		auto i = content.size();
+		total = archive_entry_size(entry);
+		content.resize(i + total);
+		size = archive_read_data(archive_obj, &content[i], total);
+
+		if (size <= 0) {
+		    error = "Libarchive was not able to extract data from " +
+			    pathname;
+		    return false;
+		}
+		content.resize(i + size);
+	    } else if (pathname == "docProps/core.xml") {
+		total = archive_entry_size(entry);
+		metadata.resize(total);
+		size = archive_read_data(archive_obj, &metadata[0], total);
+		if (size > 0) {
+		    metaxml = true;
+		    metadata.resize(size);
+		}
+	    }
+	}
+    }
+    // MSXmlParser to parse .wordprocessingml and .presentationml
+    // XlsxParser to parse .spreadsheetml
+    if (msxml) {
+	MSXmlParser xmlparser;
+	xmlparser.parse_xml(content);
+	dump = xmlparser.dump;
+    } else {
+	XlsxParser parser;
+	parser.parse(content);
+	dump = parser.dump;
+    }
+
+    // Parse if docProps/core.xml is found
+    if (metaxml)
+	parse_metadata(metadata, title, keywords, author);
```

Comment:
Calling `parse_metadata()` here sets fields `title`, `keywords`, and `author` which doesn't overlap with what's set by parsing the other file above, so there's no need to delay this parsing until here - we can just parse the data where we extract it from the zip archive above (there are several places, but you've put the code to do the parsing into a function so you can just call it in each case).

That reduces how long we need to hold that data in memory, reduces the lifetime of the `metadata` variable, and eliminates the need for the `metaxml` flag.  It'll also mean the code is easier to follow.

Reply:
Thanks for pointing this out. I have made the required changes and have updated the PR.